clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -150,15 +150,29 @@ public FilterResult filterTo(TopicPartition partition, RecordFilter filter, Byte
         return filterTo(partition, batches(), filter, destinationBuffer, maxRecordBatchSize, decompressionBufferSupplier);
     }
 
+    /**
+     * Note: This method is also used to convert the first timestamp of the batch (which is usually the timestamp of the first record)
+     * to the delete horizon of the tombstones which are present in the batch. 
+     */
     private static FilterResult filterTo(TopicPartition partition, Iterable<MutableRecordBatch> batches,
                                          RecordFilter filter, ByteBuffer destinationBuffer, int maxRecordBatchSize,
                                          BufferSupplier decompressionBufferSupplier) {
         FilterResult filterResult = new FilterResult(destinationBuffer);
         ByteBufferOutputStream bufferOutputStream = new ByteBufferOutputStream(destinationBuffer);
-
         for (MutableRecordBatch batch : batches) {
             long maxOffset = -1L;
-            BatchRetention batchRetention = filter.checkBatchRetention(batch);
+            // we first call this method here so that the flag in LogCleaner has been set
+            // which indicates if the control batch is empty or not
+            // we do this to avoid calling CleanedTransactionMetadata#onControlBatchRead
+            // more than once since each call is relatively expensive
+            filter.isControlBatchEmpty(batch);
+            long deleteHorizonMs = filter.retrieveDeleteHorizon(batch);
+            final BatchRetention batchRetention;
+            if (!batch.deleteHorizonSet())
+                batchRetention = filter.checkBatchRetention(batch, deleteHorizonMs);
```

Comment:
Since deleteHorizonMs can be obtained from batch, it's not clear why we need to pass that in as a param.

Reply:
Oh, look in comment above. This delete horizon is used for the case where we want to remove the tombstones in a single pass. On the first iteration of Log Cleaner, we are unable to remove the tombstone because no delete horizon has not been set yet. Therefore, when we compute the delete horizon, we need to pass the delete horizon back into ```checkBatchRetention``` so that tombstones can be removed in one iteration.

On second thought, I think we don't need to add an extra parameter to the ```checkBatchRetention``` method. Such logic would only need to be restricted to LogCleaner. i.e. we store the delete horizon in another variable in the Record Filter we implemented in LogCleaner.

Reply:
Alright, acknowledged. I think thats a good point.